pkg/endpointmanager/host.go

```diff
@@ -32,78 +29,30 @@ func (mgr *endpointManager) HostEndpointExists() bool {
-// from K8s.
-// This adheres to the subscriber.NodeHandler interface.
-func (mgr *endpointManager) OnAddNode(newNode *slim_corev1.Node,
-	swg *lock.StoppableWaitGroup) error {
+func (mgr *endpointManager) startNodeLabelsObserver(old map[string]string) {
+	mgr.localNodeStore.Observe(context.Background(), func(ln node.LocalNode) {
+		if maps.Equal(old, ln.Labels) {
+			return
+		}
 
-	node.SetLabels(newNode.GetLabels())
-	node.SetAnnotations(newNode.GetAnnotations())
+		mgr.updateHostEndpointLabels(old, ln.Labels)
+		old = ln.Labels
+	}, func(error) { /* Executed only when we are shutting down */ })
+}
 
+func (mgr *endpointManager) updateHostEndpointLabels(oldNodeLabels, newNodeLabels map[string]string) {
 	nodeEP := mgr.GetHostEndpoint()
 	if nodeEP == nil {
-		// if host endpoint does not exist yet, labels will be set when it'll be created.
-		return nil
-	}
-
-	newLabels := labels.Map2Labels(newNode.GetLabels(), labels.LabelSourceK8s)
-	newIdtyLabels, _ := labelsfilter.Filter(newLabels)
-	nodeEP.UpdateLabels(context.TODO(), newIdtyLabels, nil, false)
-
-	return nil
-}
-
-// OnUpdateNode implements the endpointManager's logic for reacting to updated
-// nodes in K8s.
-// This adheres to the subscriber.NodeHandler interface.
-func (mgr *endpointManager) OnUpdateNode(oldNode, newNode *slim_corev1.Node,
-	swg *lock.StoppableWaitGroup) error {
-
-	oldNodeLabels := oldNode.GetLabels()
-	newNodeLabels := newNode.GetLabels()
-
-	oldNodeAnnotations := oldNode.GetAnnotations()
-	newNodeAnnotations := newNode.GetAnnotations()
-
-	labelsEqual := comparator.MapStringEquals(oldNodeLabels, newNodeLabels)
-	annotationsEqual := comparator.MapStringEquals(oldNodeAnnotations, newNodeAnnotations)
-
-	// if labels have changed we need to recompute security ID.
-	if !labelsEqual {
-		nodeEP := mgr.GetHostEndpoint()
-		if nodeEP == nil {
-			log.Error("Host endpoint not found")
-			return nil
-		}
-		err := nodeEP.UpdateLabelsFrom(oldNodeLabels, newNodeLabels, labels.LabelSourceK8s)
-		if err != nil {
-			return err
-		}
+		log.Error("Host endpoint not found")
+		return
 	}
 
-	// Perform a SetMultiAttributes if both labels and annotations has changed,
-	// since these Set operations will wake up Observers on LocalNodeStore its
-	// better to perform one update if multiple attributes change.
-	if !labelsEqual && !annotationsEqual {
-		node.SetMultiAttributes(func(n *node.LocalNode) {
-			n.Labels = newNodeLabels
-			n.Annotations = newNodeAnnotations
-		})
-	} else if !labelsEqual {
-		node.SetLabels(newNodeLabels)
-	} else if !annotationsEqual {
-		node.SetAnnotations(newNodeAnnotations)
+	err := nodeEP.UpdateLabelsFrom(oldNodeLabels, newNodeLabels, labels.LabelSourceK8s)
+	if err != nil {
+		// An error can only occur if either the endpoint is terminating, or the
+		// old labels are not found. Both are impossible, hence there's no point
+		// in retrying.
+		log.WithError(err).Error("Unable to update host endpoint labels")
+		return
 	}
```

Comment:
It looks like this commit is only accounting for node label updates. What about node annotations?

Reply:
Sure: https://github.com/cilium/cilium/pull/29319/commits/f959d32a662f71256510a84c5c87ff346efb57c2

To give a bit of context, before these changes, the `OnUpdateNode` method of the endpointmanager was responsible for both updating the labels associated with the host endpoint and propagating label and annotation changes to the LocalNodeStore (through the `node.Set*` global setters). While the first task should be handled b the endpointmanager and has been preserved in this commit, the second is actually extraneous of this module.

For this reason, the k8s node (for the node we are running on) to LocalNode abstraction synchronization has been centralized as part of the `localNodeSynchronizer` module, which was previously responsible for the initialization only. And more specifically by the [SyncLocalNode](https://github.com/giorio94/cilium/blob/mio%2Flocal-node-refactor/daemon/cmd/local_node_sync.go#L58-L75) function which is started by the [LocalNodeStore startup hook](https://github.com/giorio94/cilium/blob/mio%2Flocal-node-refactor/pkg/node/local_node_store.go#L107). Another advantage of this approach is that now additional modules, such as WireGuard, can independently configure extra labels/annotations, automatically propagated to the CiliumNode/kvstore representation.

Reply:
>While the first task should be handled b the endpointmanager and has been preserved in this commit, the second is actually extraneous of this module.

Agreed!

>Another advantage of this approach is that now additional modules, such as WireGuard, can independently configure extra labels/annotations, automatically propagated to the CiliumNode/kvstore representation.

Now that the node annotation updates processing is (correctly) moved out of `pkg/endpointmanager`, I'm not sure if modules such as BGP are involved in annotations related updates. /cc @ldelossa Please double check in the context of https://github.com/cilium/cilium/commit/6a517a0cd7205932035b0571ce4b73be4bbe5469. (I'm not familiar with BGP side of things to make a judgement call on whether this logic is covered by tests either.)